Introduce `GraphAlgorithms.h` helper functions, and add `mlir` `include-group`

andrealinux1 opened the pull request:
- Introduce helper functions used for `clift` restructuring into the `GraphAlgorithms.h` header.
- Add the `mlir` category for the `include-group` `clang-format` directive.

aleclearmind commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -400,7 +418,11 @@ class DFSReachableState : public DFState<NodeT> {
+    auto BeginIt = llvm::GraphTraits<NodeT>::child_begin(Block);
+    auto EndIt = llvm::GraphTraits<NodeT>::child_end(Block);
+    size_t BlockSuccessorsN = std::distance(BeginIt, EndIt);
+    if ((Targets.contains(Block))
+        or ((Target == nullptr) && (BlockSuccessorsN == 0))) {
```
Drop extra parentheses around the whole `if` condition.

andrealinux1 replied:
Done, this part of the code is now in `revng-c`.

aleclearmind commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+      bool AreEquivalent = equal(*RegionIt1, *RegionIt2);
+      if (Intersects
+          and (((!IsIncluded) and (!IsIncludedReverse)) or AreEquivalent)) {
+        (*RegionIt1).insert((*RegionIt2).begin(), (*RegionIt2).end());
```
`(*x).insert` -> `x->insert`

andrealinux1 replied:
Right, changed in 7b60c49bba5376fbe0c3de3d0f15e2c1c810e9c7.

aleclearmind commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+      if (Intersects
+          and (((!IsIncluded) and (!IsIncludedReverse)) or AreEquivalent)) {
+        (*RegionIt1).insert((*RegionIt2).begin(), (*RegionIt2).end());
+        R.erase(RegionIt2);
```
Are you sure this is safe?

If you need to iterate on a container and erase certain elements at the same time usually you do

```c++
for (auto It = V.begin(); It != V.end(); /* Nothing here */) {
  if (WantToDelete) {
    It = V.erase(It);
  } else {
    ++It;
  }
}
```

andrealinux1 replied:
You are right in the general case, but in this particular implementation an `erase` operation on the container is always followed by a `return` from the current function, for algorithm-related reasons. So in the end, every time we have an `erase`, we conclude our visit on the container and start again. Therefore I think that the implementation is correct.

andrealinux1 replied:
Right, I added a comment specifying the situation in the code now moved in `revng-c`.

aleclearmind commented on tests/unit/GraphAlgorithms.cpp:
```diff
@@ -178,9 +224,285 @@ BOOST_AUTO_TEST_CASE(GetBackedgesTest) {
+  using BlockSet = SmallSetVector<NodeType *>;
+  using BlockSetVect = llvm::SmallVector<BlockSet>;
+
+  // Compute the backedges set.
```
No dots at the end of comments

andrealinux1 replied:
Regarding this point, since I saw that scattered in the codebase there are different behaviors about this. What should I do with explanation comments that have more than one sentence? like:
```
// Roses are red.
// Violets are blue.
// Fer are faster, Merc one-two
```
or
```
// Roses are red.
// Violets are blue.
// Fer is faster, Merc one-two.
```
?

aleclearmind replied:
No dots at the end of comments for single line comments only. For the rest, the second option you gave it's OK.

andrealinux1 replied:
Right, in the end i went with no dots for _single sentence_ (not _single line_) comments, which seemed more coherent with the rest of the codebase (even if there are some mixed uses). Multi sentence comments have dots.

aleclearmind commented:
Drop the `clift` prefix from commit messages. `revng` doesn't know anything about clift.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -537,3 +537,15 @@ nodesBetweenNewReverse(GraphT Source,
 }
+
+template<class GraphT>
+bool isDAG(GraphT Graph) {
```
I changed this in 7b60c49bba5376fbe0c3de3d0f15e2c1c810e9c7, but I have a question on this point. In this PR, `GraphT` is used with actual type always as a pointer (e.g., `mlir::Block *`). Is it good-practice to use `const GraphT &` as signature in this case (it may result confusing from an user point of view), or should we do something different like restricting the input template type with something like `requires std::is_pointer_v<GraphT>`?

aleclearmind replied:
You're right. Using `const GraphT &` is a bad choice. Let's revert back to just `GraphT`.

One day we'll have the concept `HasGraphTraits`, where we can also add something to assert that the size is `<= sizeof(intptr_t)`.

andrealinux1 replied:
One day! Reverted back to the old situation in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd.

andrealinux1 commented on tests/unit/GraphAlgorithms.cpp:
```diff
@@ -178,9 +224,285 @@ BOOST_AUTO_TEST_CASE(GetBackedgesTest) {
+                                           ShortestPathFromEntry,
+                                           RPOT);
+
+  llvm::dbgs() << "Elected entry: " << Entry->getIndex() << "\n";
```
Right, this was a leftover from the in-tree implementation.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -290,10 +293,21 @@ nodesBetweenReverse(G Source,
+using SmallMapVector = llvm::SmallMapVector<K, V, 4>;
+
+template<typename K, typename V>
+using SmallDenseMap = llvm::SmallDenseMap<K, V, 4>;
```
I moved this into the `revng::detail` namespace in `7b60c49bba5376fbe0c3de3d0f15e2c1c810e9c7`. A practical problem: all these namespaces. in conjuction with template parameters make signatures very long, so I thought of creating an alias namespace for `revng::detail::` called `dt::` to make code slimmer. Is this viable?

aleclearmind replied:
I'd push these `using`s inside the classes that use them.
You can also use the expanded version if not possible, or you can add a template parameter defaulted to `4` for the small size.

For instance, `SmallDenseMap` can become a using (private?) in `DFSReachableState`

> alias namespace for revng::detail:: called dt:: to make code slimmer

Nah.

andrealinux1 replied:
Ok, in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd I removed the 3 using, and pushed them into the classes that used them in some cases, and just used the expanded version where this was not possible (e.g., on free function signatures).

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -400,7 +418,11 @@ class DFSReachableState : public DFState<NodeT> {
     // `Targets` set.
-    if (Targets.contains(Block)) {
+    auto BeginIt = llvm::GraphTraits<NodeT>::child_begin(Block);
+    auto EndIt = llvm::GraphTraits<NodeT>::child_end(Block);
```
Right, changed in 7b60c49bba5376fbe0c3de3d0f15e2c1c810e9c7.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+}
+
+template<class NodeT>
+bool simplifyRegionsStep(llvm::SmallVector<SmallSetVector<NodeT>> &R) {
```
As we discussed, I moved a lot of the code previously part of this PR closer to the clift implementation in `revng-c`.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+      revng_assert(!Children.empty());
+      return &Children[0];
+    }
+    __builtin_unreachable();
```
Right, changed in 7b60c49bba5376fbe0c3de3d0f15e2c1c810e9c7.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+template<class GraphT>
+auto successor_range(GraphT Block) {
+  return child_range(Block);
+}
```
- I have both just to fold code, `child_range` accepts the template parameter, so that I can call it both for `successor_range` and `predecessor_range` just with a different template `GT` parameter to do the right thing and call `llvm::make_range` once. I can of course duplicate the call to `llvm::make_range` in both `successor_range` and `predecessor_range`
- I removed the `_range` suffix, but unfortunately I could not simply use the `successors` and `predecessors` name due to conflict with the `successors` and `predecessors` defined for `llvm::BasicBlock` in `llvm/include/llvm/IR/CFG.h`. I used `blockSuccessors` and the like, otherwise I can wrap inside a namespace I guess. I made this changes in bd8b633422434d6d23731090767b287409f236f1 to avoid changing references from previous comments, this will be squashed OC.

aleclearmind replied:
1. ACK, but `successor_range` is exactly the same as `child_range` without specifying the default template argument. Collapse them.
2. Uhm, not a fan of `blockSuccessors`. These work for all the guys implementing graph traits. Maybe `graph_successors`?

andrealinux1 replied:
1. Ah ok right, I did them in the previous way to avoid confusion on successor/predecessor, but after thinking about it collasping them seems reasonable, done in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd.
2. Used `graph_successors` as suggested in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+}
+
+template<class GraphT>
+GraphT get_successor(GraphT Block, ptrdiff_t I) {
```
1. I changed this in bd8b633422434d6d23731090767b287409f236f1.
2. The point of having the `getBlockSuccessor` laid out like this is mainly in order to be compatible with all the template based infrastructure. This seems to me the only way of requiring only `GraphTraits` implemented on `GraphT` and be able to extract the n-th successors. Are you suggesting to always materialize the list of successors on the call-site side of the thing?

aleclearmind replied:
2. I think `get_successor` is a "dangerous" API to offer, since it might lead you into writing something like
   ```c++
   for (unsigned I = 0; I < ...; ++I)
     doSomething(getSuccessor(Graph, I));
   ```
   So, yes, I'm suggesting not to offer this API and let the user do this by hand if he really really needs to.

andrealinux1 replied:
Ah ok, now your point is clear. I removed the function in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd, and I materialize this when needed.

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+template<class NodeT>
+bool setContains(SmallSetVector<NodeT> &Set, NodeT &Element) {
+  return Set.contains(Element);
+}
```
Ok, I removed it in 1766e1f5f0031d72f59ba843709c7039ba64aa8c.

aleclearmind commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+bool equal(SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+  return llvm::set_is_subset(First, Second)
+         and llvm::set_is_subset(Second, First);
+}
```
OK, but those are not `SmallSetVector`-specific.

I see no point in having a `subset` function which is just a wrapper around `set_is_subset`, just use `set_is_subset` directly.
I see the point for `equal` and `intersect`, but then do as `SetOperations.h` does (in fact, we can create our own `include/revng/ADT/SetOperations.h`) and make those function generic (i.e., not `SmallSetVector`-specific).

Additional note: `intersect` is a verb representing an action but the function actually returns a `bool`. Rename to `intersects`.

andrealinux1 replied:
You are right. In ab8422012f2a7aca302f31efb79cf14a80453de6 I added the new `include/revng/ADT/SetOperations.h` header, and defined `equal` and `intersect`, while I dropped `set_is_subset`.

The only thing is that, due to the internal functioning of `llvm::set_intersect` (there is some idiosyncrasy with how `llvm` handles the internal `erase` operation with `SmallSet`s like type) , there were some problems in adopting the template signature for `SetTy`. I think that the easiest solution is to make use of a new `llvm::set_intersection` function defined recently in the LLVM codebase, which does exactly what we want.

Therefore, I cherry-picked in https://github.com/revng/llvm-project/pull/28/commits/fd9d91de43f62a74bc9ba4e297389b7392adcad3, in the sister PR to the `llvm-project` repo, the, well isolated, llvm upstream commit that introduce it, so that we can adopt this approach without having to make hacks. Let me know if this seems reasonable to you.

aleclearmind commented on tests/unit/GraphAlgorithms.cpp:
```diff
@@ -178,9 +111,43 @@ BOOST_AUTO_TEST_CASE(GetBackedgesTest) {
+  revng_check(Backedges[1].first == NLG.SecondLoopLatch);
+  revng_check(Backedges[1].second == NLG.LoopHeader);
+
+  BlockSetVect Regions;
+  for (EdgeDescriptor Backedge : Backedges) {
+    BlockSet RegionNodes = nodesBetweenNew(Backedge.second, Backedge.first);
+    Regions.push_back(std::move(RegionNodes));
+  }
+
+  revng_check(Regions.size() == 2);
+  revng_check(Regions[0][0] == NLG.LoopLatch);
+  revng_check(Regions[0][1] == NLG.SecondLoopHeader);
+  revng_check(Regions[1][0] == NLG.SecondLoopLatch);
+  revng_check(Regions[1][1] == NLG.LoopHeader);
+  revng_check(Regions[1][2] == NLG.SecondLoopHeader);
+  revng_check(Regions[1][3] == NLG.LoopLatch);
```
* Let's not use the term regions
* Add a comment saying that we want to test that nodes reachable from the extremes of the a backedge is basically the whole loop

andrealinux1 replied:
In b39eb194c7c8c536266997effafe3b5ec39dd6bb I removed the _region_ terminology, going for a more neutral _loop_, since here we do not have the _region_ concept.

I also added a comment explaining what's the purpose of the above tests.

andrealinux1 commented:
I addressed the following problems that we identified in the previous state of the MR:
- We changed the terminology used for the classes representing the DFS internal state. We moved away from `DFState`, `DFSBackedgeState` in favour of `DFStackReachable` and the like, in order to better convey the idea of the internal state of the visit stack.
- Regarding the aforementioned classes, we also switched from the use of inheritance, for implementing the `DFStackBackedge` and `DFStackReachable` classes (used as the internal state in the DFS algorithms), to the use of composition.
- Removed various useless fields from the aforementioned classes, like the internal storage of `Backedges` etc.
- We discussed the possibility of removing the `std::function IsValid` field, I opened a specific comment on that.
- Added a TODO suggesting to remove the `IgnoreList` parameter from the `nodesBetween` implementation, once the use in AVI is dropped or modified.
- Added a TODO suggesting to drop the `IsValid`  parameter once the last use in `revng-c` using it is dropped with the introduction of MLIR.

andrealinux1 commented on include/revng/BasicAnalyses/AdvancedValueInfo.h:
```diff
@@ -627,18 +627,22 @@ class Expression {
+      // `LimitedStartBB` is manually inserted as the first element in
+      // `ReachableVector`, so we do not need to insert it in the vector
+      if (BB != LimitedStartBB)
+        ReachableVector.push_back(BB);
```
Fixed in 626415ff79c8c9cbfe2cdf725adb03cf4074c12a

andrealinux1 commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -154,286 +158,200 @@ entryPoints(GraphType &&Graph) {
+  //       one use-case, the number of valid nodes is very small, while the
+  //       invalid nodes is usually very high (the rest of the graph), therefore
+  //       the underlying set could become very huge.
+  const std::function<bool(const NodeT)> IsValid;
```
According to what we discussed, I've dropped the `std::function` field, and the `getBackedges`, `nodesBetween` and `nodesBetweenReverse` now accept a template parameter providing either a `BlackList` or a `WhiteList` of nodes. Let me know what you think about the code.

aleclearmind commented on include/revng/Support/GraphAlgorithms.h:
```diff
@@ -443,42 +514,62 @@ getBackedges(GraphT Block,
-    for (GraphT Element : *IgnoreList) {
+  // it contains
+  if (IgnoreL != nullptr) {
+    for (GraphT Element : *IgnoreL) {
```
No abbreviations

andrealinux1 replied:
I guess this was due to an outdated commit, the filter parameter is now called `Set` for the `nodesBetweenImpl` function and friends, and this change has been introduced when dropping the old `nodesBetween` in 510afd77cd288a2f899ef0afaa53590047c584c2.

aleclearmind commented on include/revng/BasicAnalyses/AdvancedValueInfo.h:
```diff
@@ -627,18 +627,29 @@ class Expression {
+    // We want to stop our backward visit at the `StopAt` node, so we actually
+    // need to blacklist the predecessors of `StopAt`
+    SmallPtrSet<BasicBlock *, 4> IgnoreList;
+    for (BasicBlock *Pred : predecessors(StopAt)) {
+      IgnoreList.insert(Pred);
+    }
```
Consider using `llvm::copy(predecessors(StopAt), std::inserter(IgnoreList.begin(), IgnoreList)`

andrealinux1 replied:
Now in 1d384e05daf8ce38e943a63a3267d8a72118b267 we make use of `llvm::copy` and `llvm::copy_if` a couple of lines below.

aleclearmind commented:
* "GraphAlgorithms: introduce helpers used in clift ": speak in more broad terms, clift is not something known to a revng dev
* "GraphAlgorithms: drop old nodesBetween impl":
  * No abbreviations
  * Elaborate on why we're now able to drop it

andrealinux1 commented:
> * "GraphAlgorithms: introduce helpers used in clift ": speak in more broad terms, clift is not something known to a revng dev
> 
>     * "GraphAlgorithms: drop old nodesBetween impl":
>       
>       * No abbreviations
>       * Elaborate on why we're now able to drop it

Right, I changed the message in order to not refer to clift details, and give a brief but descriptive idea of the changes introduced.
